Validate EDF by counting BallotStyles

stratofax opened the pull request:
Validate a JSON EDF by counting BallotStyles. Invoke from the command line as follows:

```
ballotmaker validate --edf ~/june_test_case.json
```

100% test coverage!

Although this tests to ensure the specified EDF file exists, any JSON errors will trigger an exception.

ion-oset commented on src/electos/ballotmaker/validate_edf.py:
```diff
@@ -1,6 +1,8 @@
-from electos.ballotmaker.constants import NO_ERRORS, NO_FILE
+from electos.ballotmaker.constants import NO_DATA, NO_ERRORS, NO_FILE
+from electos.ballotmaker.read_edf import read_edf
+from genericpath import isfile
```
`genericpath` is a stdlib module: move it up to be grouped with `from pathlib...`

`isort` will take care of this when you get around to it.

stratofax replied:
Good catch! VSCode threw in `from genericpath import isfile` and I didn't think too much about it at the time, although `isort` is placing it where it doesn't belong, too.

In fact, I prefer `Path.is_file()` because it's one less import. I made this change globally. Still don't know why isort put that at the bottom of the import list, but not an issue now.

ion-oset commented on tests/test_validate_edf.py:
```diff
@@ -1,12 +1,28 @@
 from electos.ballotmaker import validate_edf
-from electos.ballotmaker.constants import NO_ERRORS, NO_FILE
+from electos.ballotmaker.constants import NO_DATA, NO_ERRORS, NO_FILE
+from genericpath import isfile
```
👆🏼 See above about `genericpath`.

stratofax replied:
Fixed as noted above.

ion-oset commented on tests/test_validate_edf.py:
```diff
@@ -1,12 +1,28 @@
+empty_file = Path("empty.json")
+empty_file_path = Path(test_dir, empty_file)
+test_file = Path("june_test_case.json")
+full_test_path = Path(test_dir, test_file)
```
If these are constants, capitalize them,

stratofax replied:
I'm not sure if they're constants, yet. I could see changing the values, especially if I used test fixtures. Also, in my mind, constants are things like strings and ints, not more obscure things like Paths -- not a string. Maybe I'm being old-fashioned. I think this one merits further discussion. I'm going to leave as-is for now.

ion-oset reviewed (changes requested):
Mostly style feedback. Otherwise LGTM.

stratofax commented:
I removed some outdated code from `versadm/ballots` that used `os` instead of `pathlib` and was also breaking my GitHub Actions. When I'm done with this round of feature development, the contents of `versadm/ballots` will be integrated into `ballotmaker` or removed.

stratofax commented:
@ion-oset I made the changes you suggested, except CAPITALIZING -- I'd like to defer that decision until after we've had a chance to discuss what should be considered a constant, because that's not really a thing in Python, but it's a convention we should agree upon. Please let me know if you approve the changes as they stand.

trustthevote commented:
Sounds good, thanks

ion-oset commented on tests/test_cli.py:
```diff
@@ -1,4 +1,5 @@
 from pathlib import Path
+from unittest import result
```
This isn't being used. All instances of `result` below are variables returned by `runner.invoke`.

stratofax replied:
I think this line was inserted by VSCode, which I'm not happy about. I need to figure out how to turn this "feature" off. The same thing happened with `isfile`. Maybe there's such a thing as too much automation...

stratofax replied:
I removed this import line

ion-oset replied:
> I think this line was inserted by VSCode, which I'm not happy about. 

Ugh. That's annoying, all the more so because it keeps doing it ☹️ 

> Maybe there's such a thing as too much automation...

😝

ion-oset commented on src/electos/ballotmaker/validate_edf.py:
```diff
@@ -0,0 +1,22 @@
+    # is the EDF a file?
+    if _edf is None:
+        return NO_FILE
+    if Path.is_file(_edf) is False:
```
```suggestion
    if not _edf.is_file():
```
- `is_file` is an instance method, not a class method. (What you did works here but it would go wrong if inheritance was involved.)
- Per PEP-8, don't use `is False` on boolean tests.

stratofax replied:
Agreed and done.

ion-oset commented on tests/test_read_edf.py:
```diff
@@ -0,0 +1,15 @@
+
+def test_read_edf():
+    # is the test file available?
+    assert Path.is_file(full_test_path)
```
As above: `Path.is_file` is an instance method not a class method:

```suggestion
    assert full_test_path.is_file()
```

stratofax replied:
Fixed

ion-oset commented on tests/test_read_edf.py:
```diff
@@ -0,0 +1,15 @@
+from pathlib import Path
+
+from electos.ballotmaker import read_edf
```
In the source files you are using this instead. I think the form which imports the function directly is more readable, but I don't think it matters which form you choose as long as it's consistent, especially since the module and function names are the same.

```suggestion
from electos.ballotmaker.read_edf import read_edf
```

stratofax replied:
Agreed and updated.

ion-oset commented on tests/test_validate_edf.py:
```diff
@@ -0,0 +1,27 @@
+from pathlib import Path
+
+from electos.ballotmaker import validate_edf
```
👆🏼 Ditto for comment about importing `read_edf` above.

stratofax replied:
Agreed and fixed

ion-oset commented on tests/test_validate_edf.py:
```diff
@@ -0,0 +1,27 @@
+    # assert make_ballots.make_ballots(_edf=None) == NO_FILE
+    assert validate_edf.validate_edf(_edf=None) == NO_FILE
+    # ensure not_a_file is actually not a file
+    assert Path.is_file(not_a_file) == False
```
☝🏼 Ditto for comment above about `Path.is_file`

stratofax replied:
Fixed.

ion-oset commented on tests/test_validate_edf.py:
```diff
@@ -0,0 +1,27 @@
+    # assert Path.is_file(empty_file_path)
+    # assert validate_edf.validate_edf(empty_file_path) == NO_DATA
+
+    assert Path.is_file(full_test_path)
```
☝🏼 Ditto for comment above about `Path.is_file`

stratofax replied:
Fixed,

ion-oset reviewed (changes requested):
Minor fixes and style changes.

stratofax commented:
@ion-oset I updated code based on your comments from today.

ion-oset commented on tests/test_validate_edf.py:
```diff
@@ -15,13 +15,10 @@
-    # test empty json file
-    # assert Path.is_file(empty_file_path)
-    # assert validate_edf.validate_edf(empty_file_path) == NO_DATA
+    assert not_a_file.is_file() == False
```
```suggestion
    assert not not_a_file.is_file()
```

stratofax replied:
The double negative was too much for my brain, so I changed it like this:

```
assert not imaginary_file.is_file()
```

stratofax commented:
I renamed `not_a_file.json` to `imaginary_file.json` to avoid a double negative. All your specified changes are complete. I've already started updating `make_ballots.py` and the associated test file, but I'd like to submit those changes in a separate pull request, with some logging code I added this morning.

stratofax commented:
I'm going to merge this branch now onto the TTV repo, because I've got another PR in the wings, and I'm already looking at some file merges. @ion-oset your reviews and comments have been incredibly helpful; I'm going to proactively apply what we've covered here to work on the other branch as well.
